tests/codex/stores/testqueryiterhelper.nim

```diff
@@ -0,0 +1,65 @@
+import std/sugar
+
+import pkg/stew/results
+import pkg/questionable
+import pkg/chronos
+import pkg/datastore/typedds
+import pkg/datastore/sql/sqliteds
+import pkg/codex/stores/queryiterhelper
+import pkg/codex/utils/asynciter
+
+import ../../asynctest
+import ../helpers
+
+proc encode(s: string): seq[byte] =
+  s.toBytes()
+
+proc decode(T: type string, bytes: seq[byte]): ?!T =
+  success(string.fromBytes(bytes))
+
+asyncchecksuite "Test QueryIter helper":
+  var
+    tds: TypedDatastore
+
+  setupAll:
+    tds = TypedDatastore.init(SQLiteDatastore.new(Memory).tryGet())
+
+  teardownAll:
+    (await tds.close()).tryGet
+
+  test "Should auto-dispose when QueryIter finishes":
```

Comment:
If I'm reading this right, this test runs the iter to completion. It'd be really nice to have a test that does not. So it iterates only the first few items of a longer list.

Reply:
Not sure what such test would suppose to check?

Reply:
I believe @benbierens wants to guarantee that the query iterator is disposed even if it's not run to completion (which, I think, currently it won't be).

Reply:
Ok, not sure if that's possible.

Reply:
Ah yes!
You see, if you use the leveldb helper, then in the setup you create the DB and in teardown you dispose it. If your code does not correctly release the iterators, then your teardown will throw. This way you know there's a problem. (Sqlite doesn't do this, it quietly ignores undisposed iterators.)
So then simply having a test that partially iterates a list would reveal the problem if it's there.

Reply:
>So then simply having a test that partially iterates a list would reveal the problem if it's there.

I mean we know that the iterator will not be disposed in this case. If I would add such test it would always fail on the teardown. Also such test would be impossible to fix currently according to the `TODO` mentioned in [this comment](https://github.com/codex-storage/nim-codex/pull/811#discussion_r1627262366).

We could try to assert that iter isn't disposed in such scenario, but that's even worse, because we would specify (tests are specs) that not disposing an iter is expected behavior. So I prefer to leave it unspecified.

If I'm missing something please provide me an example how such test should look like.

Reply:
It sounds to me like using the query-iter-helper will guarantee we leak the iterator whenever a query is not run to completion. This can be solved by the other GC in the future. But I suppose in the meantime, we need a way to manually dispose the iter when the user of the iter knows that it's not going to complete.
Either asyncIter has to support manually disposing (for now), or we shouldn't use it instead of queryIter. :| This sucks, but crashes due to leaks or other memory issues are a nightmare to even detect.

Reply:
Please elaborate on how using this module will guarantee that "we leak the iterator", maybe an example?